### ISSUE TITLE
hdf5 model file for r941_min_sup_g507?

### ISSUE THREAD

**Issue by gallardo-seq** [nanoporetech/medaka#308]
Hi Medaka team, I was wondering if there's a way obtain an hdf5 file for the new r941_min_sup_g507 model included in the new v 1.4.1 release? I have successfully used custom hdf5 model files for our MrHAMER pipeline, but it seems the recent models are now included in a tar.gz file (which does not contain an hdf5 file inside). Unfortunately, our MrHAMER pipeline is currently set up with medaka v 1.0.1, so simply updating to the latest version would not work for us without a major overhaul of our code and testing. Are there any scripts within the Medaka package that would convert tar.gz models to hdf5? I would really appreciate any information or help with this.

**Comment by cjw85** [comment]
We do not have a script to convert the tar.gz files to HDF (these files are a bundling of Tensorflow's model format rather than the Keras specific HDFs). We will add one if we find some time, but can't promise a timeline.

Two additional points. I can't see any barriers to updating from v1.0.1 to v1.4.1, the CLI has not changed neither has the fundamental methodolgy. You can review the [CHANGELOG](https://github.com/nanoporetech/medaka/blob/master/CHANGELOG.md) to see what's changed. In relation to the workflow of MrHAMER, you might find the `medaka smolecule` command useful which was implemented to more efficiently handle the use case of processing many small independent consensus calculations.

**Comment by gallardo-seq** [comment]
Thanks for the quick note. That's a very helpful explanation, I'll start looking on my end for any scripts that can convert the TF model to Keras HDF. If you end up finding some time to convert these to hdf, I'd really appreciate it if you can keep me in the loop (even if not in an official release).

Regarding the update from version 1.0.1 to 1.4.1, somewhere along the way there was a change in medaka stitch, which now requires a DRAFT variable. I did a pilot run with v1.4.1 where I provided the racon draft sequence, and medaka stitch either outputs an empty fasta file, or outputs the racon sequence verbatim. I will look into it more in the next few weeks, and will keep you posted once I have more specifics.

Thanks for the heads up about medaka smolecule. I did try out an earlier smolecule release with the --method racon option, and it is very much quicker than our MrHAMER pipeline wrapper. The only thing is that our MrHAMER pipeline is reference-guided, whereas the smolecule approach seems to be novo, , which results in some slight dips in overall accuracy for our HIV samples. Maybe this is something that the new g507 models will help with. Will try this next week and keep you posted.

Thanks again!

**Comment by cjw85** [comment]
>  I did a pilot run with v1.4.1 where I provided the racon draft sequence, and medaka stitch either outputs an empty fasta file, or outputs the racon sequence verbatim.

This would indicate that the bam file contained no alignments for medaka to process. By default (at the request of many users) medaka fills 0-depth regions with the input draft sequence. The `--no-fillgaps` disables this behaviour.

With respect to the smolecule code, in principle you could adapt it to accept a reference sequence rather than being denovo. The entry poitn [here](https://github.com/nanoporetech/medaka/blob/master/medaka/smolecule.py#L390) does only a few things:

1) parse input data
2) form POA consensus and return alignments of reads to consensus
3) writes a BAM with all data
4) runs medaka's consensus

The trick to this being fast is that the last step wants to operate on a single BAM file, not one per read cluster. In fact if you edited [this function](https://github.com/nanoporetech/medaka/blob/master/medaka/smolecule.py#L336) to skip the POA step, and ensured you gave the program a FASTA where the first "read" in each cluster were the corresponding reference, I believe you will get the reference-based approach you desire. 

On a side note the standard consensus models in medaka shouldn't be used when reads are aligned to a reference, the analagous models named "variant" are perhaps more appropriate since they are trained for this task. Its worth playing with both.

**Comment by gallardo-seq** [comment]
I see, that makes sense. Does that mean that if I use the --no-fillgaps option that the DRAFT argument wouldn't be required?

Thanks for the detailed notes regarding the smolecule code! I will take a closer look at lines #390 and #336 and see if there's a way to adapt the code for our specific purpose. It's good to know that the single BAM file is the way to go, that explains the slower performance we had with out MrHAMER wrapper.

Regarding the consensus models vs the variant model, I was not aware of this nuance. I will take a look at our control datasets and see if the variant models can push our accuracy up. Will report back on these points next week.

Thanks!

**Comment by cjw85** [comment]
Unfortunately the DRAFT argument is required even with the `--no-fillgaps` option. If should always be given as the same file used as a reference used for alignment to create the input `.bam` for `medaka consensus`.

Thinking more about MrHAMER: the Read class within the smolecule code was written to allow various workflows; you should be able to implement any combination of SPOA- or reference-based approach (or combination thereof) that you want. Again just amending the `_read_worker` function is all you need to do to experiment.
